### PR TITLE
SALTO-4094: Create priority Instance For Policy

### PR DESCRIPTION
**Pull request by idoamit8**
In this PR I added priority instance for policy

---

_Additional context for reviewer_
1. I modified the code that supports priority policy rules filter to support policies as well. 

---
_Release Notes_: 
_Okta Adapter:_
* New instances responsible for policies priorities were added. 

---
_User Notifications_: 
_Okta Adapter:_
* New instances responsible for policies priorities were added.

### CODE REVIEW

File: packages/okta-adapter/src/constants.ts [hunk 1]
```diff
@@ -51,6 +51,13 @@ export const IDP_RULE_PRIORITY_TYPE_NAME = 'IdentityProviderPolicyRulePriority'
 export const MFA_RULE_PRIORITY_TYPE_NAME = 'MultifactorEnrollmentPolicyRulePriority'
 export const SIGN_ON_RULE_PRIORITY_TYPE_NAME = 'OktaSignOnPolicyRulePriority'
 export const PASSWORD_RULE_PRIORITY_TYPE_NAME = 'PasswordPolicyRulePriority'
+export const ACCESS_POLICY_PRIORITY_TYPE_NAME = 'AccessPolicyPriority'
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/constants.ts:54]: as far as I remember there's no priorities in the policy level for `AccessPolicy` (and also to some other policy types), is that changed?

File: packages/okta-adapter/src/constants.ts [hunk 2]
```diff
@@ -80,6 +87,7 @@ export const POLICY_RULE_PRIORITY_TYPE_NAMES = [
   PROFILE_ENROLLMENT_RULE_PRIORITY_TYPE_NAME,
   AUTHORIZATION_POLICY_RULE_PRIORITY_TYPE_NAME,
 ]
+export const POLICY_PRIORITY_TYPE_NAMES = [SIGN_ON_POLICY_PRIORITY_TYPE_NAME]
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/constants.ts:90]: is this intentional ?

File: packages/okta-adapter/src/config.ts [hunk 3]
```diff
@@ -303,17 +304,28 @@ const getPolicyConfig = (): OktaSwaggerApiConfig['types'] => {
   })
   return Object.assign({}, ...policiesConfig)
 }
-const getPoliceyRulePriorityConfig = (): OktaSwaggerApiConfig['types'] => {
+const getPoliceyAndPolicyRulePriorityConfig = (): OktaSwaggerApiConfig['types'] => {
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/config.ts:307]: missed earlier 😅 
```suggestion
const getPolicyAndPolicyRulePriorityConfig = (): OktaSwaggerApiConfig['types'] => {
```

File: packages/okta-adapter/src/config.ts [hunk 4]
```diff
@@ -303,17 +304,28 @@ const getPolicyConfig = (): OktaSwaggerApiConfig['types'] => {
...
     [typeName]: {
       deployRequests: {
         add: { url: '', method: 'put' },
         modify: { url: '', method: 'put' },
+        remove: { url: '', method: 'delete' },
       },
     },
   }))
-  return Object.assign({}, ...policyRulePrioritiesConfig)
+  const policyPrioritiesConfig = POLICY_PRIORITY_TYPE_NAMES.map(typeName => ({
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/config.ts:318]: is `policyPrioritiesConfig` different in anyway from `policyRulePrioritiesConfig` ? can we consolidate ?

File: packages/okta-adapter/src/filters/policy_priority.ts [hunk 5]
```diff
@@ -129,42 +158,38 @@ const getParentPolicy = (rule: InstanceElement): InstanceElement | undefined =>
...
- * for the policy rules' priorities. Each priority instance contains the policy rules sorted
- * by their priority, including the default rule. The default rule is always set to be last.
- * In deployment, we deploy the priorities, not the policy rules themselves.
+ * Manages the priorities of policies and policy rules by generating an InstanceElement
+ * for the priorities. Each priority instance contains the instances sorted
+ * by their priority, including the default instance. The default instance is always set to be last.
+ * In deployment, we deploy the priorities, not the instances themselves.
  */
 const filter: FilterCreator = ({ config, client }) => ({
   name: 'policyRulePrioritiesFilter',
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/filters/policy_priority.ts:195]: nit: align this as well

File: packages/okta-adapter/src/filters/policy_priority.ts [hunk 6]
```diff
@@ -173,6 +198,12 @@ const filter: FilterCreator = ({ config, client }) => ({
     const policiesRules = instances.filter(instance =>
       ALL_SUPPORTED_POLICY_RULE_NAMES.includes(instance.elemID.typeName),
     )
+    const priorityTypes = POLICY_RULE_PRIORITY_TYPE_NAMES.map(name => createPriorityType(name, 'defaultRule')).concat(
+      POLICY_PRIORITY_TYPE_NAMES.map(name => createPriorityType(name, 'defaultPolicy')),
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/filters/policy_priority.ts:202]: does all supported policies has the concept of `defaultPolicy` ?

File: packages/okta-adapter/src/filters/policy_priority.ts [hunk 7]
```diff
@@ -73,22 +76,48 @@ const POLICY_NAME_TO_PRIORITY_NAME: Record<string, string> = {
...
     elemID: new ElemID(OKTA, typeName),
     fields: {
       priorities: {
         refType: new ListType(BuiltinTypes.NUMBER),
         annotations: { [CORE_ANNOTATIONS.CREATABLE]: true, [CORE_ANNOTATIONS.UPDATABLE]: true },
       },
-      defaultValue: {
+      [defaultFieldName]: {
         refType: BuiltinTypes.STRING,
         annotations: { [CORE_ANNOTATIONS.CREATABLE]: true, [CORE_ANNOTATIONS.UPDATABLE]: true },
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/filters/policy_priority.ts:92]: sorry it's on the previous PR, but isn't the default shouldn't be `UPDATABLE`?

File: packages/okta-adapter/src/filters/policy_priority.ts [hunk 8]
```diff
@@ -114,9 +143,9 @@ const createPolicyRulePriorityInstance = ({
   })
 }
 
-// For AccessPolicy, the priority index starts from 1, while for others it starts from 0.
-const setPriority = (typeName: string, priority: number): number => {
-  if (typeName === ACCESS_POLICY_TYPE_NAME) {
+// For AccessPolicyRules, the priority index starts from 0, while for others it starts from 1.
+export const setPriority = (typeName: string, priority: number): number => {
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/filters/policy_priority.ts:147]: probably leftover from the older version :) same for `deployPriorityChange`

```suggestion
const setPriority = (typeName: string, priority: number): number => {
```

File: packages/okta-adapter/src/filters/policy_priority.ts [hunk 9]
```diff
@@ -196,52 +227,71 @@ const filter: FilterCreator = ({ config, client }) => ({
...
+            if (positionsBefore[priority]?.elemID.getFullName() !== ref.elemID.getFullName()) {
               await deployPriorityChange({
                 client,
                 priority,
-                rule: rule.value,
+                instance: ref.value,
                 config,
+                additionalUrlVars: ALL_SUPPORTED_POLICY_RULE_NAMES.includes(ref.value.elemID.typeName)
+                  ? { ruleId: ref.value.value.id, policyId: getParentPolicy(ref.value)?.value.id }
+                  : { policyId: ref.value.value.id },
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/filters/policy_priority.ts:294]: please extract to function

File: packages/okta-adapter/test/filters/policy_priority.test.ts
Comment (shir-reifenberg) [inline packages/okta-adapter/test/filters/policy_priority.test.ts:1]: I think we should probably cover all supported policies and rules, cause not all behaves the same
If you use `.each` is shouldn't be much effort as well - https://jestjs.io/docs/api#each

File: packages/okta-adapter/src/weak_references/policy_priorities.ts [hunk 10]
```diff
@@ -98,13 +102,13 @@ const removeMissingPolicyRulePriorities: WeakReferencesHandler['removeWeakRefere
     const errors = fixedElements.map(instance => ({
       elemID: instance.elemID.createNestedID('priorities'),
       severity: 'Info' as const,
-      message: `Deploying ${instance.elemID.typeName} without all attached priorities for rules.`,
-      detailedMessage: `This ${instance.elemID.typeName} is attached to some rules that do not exist in the target environment. It will be deployed without referencing these.`,
+      message: `Deploying ${instance.elemID.typeName} without all attached priorities.`,
+      detailedMessage: `This ${instance.elemID.typeName} is attached to some instances that do not exist in the target environment. It will be deployed without referencing these.`,
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/weak_references/policy_priorities.ts:106]: let's keep use `rules` and `policies` instead `instances` (same for `message`)

**Review by shir-reifenberg — COMMENTED** [review]
Looks good! added some questions mainly around what policies should be covered here
also, please adjust release notes (this also solves the bug assigned to you, right ?)

File: packages/okta-adapter/src/constants.ts (hunk 1 above)
Comment (idoamit8) [reply packages/okta-adapter/src/constants.ts:54]: Yes, its leftover

File: packages/okta-adapter/test/filters/policy_priority.test.ts
Comment (idoamit8) [reply packages/okta-adapter/test/filters/policy_priority.test.ts:1]: Done 😎

File: packages/okta-adapter/src/weak_references/policy_priorities.ts [hunk 11]
```diff
@@ -23,53 +23,59 @@ import {
...
 const log = logger(module)
+const getInstanceAttribute = (instance: InstanceElement): string =>
+  POLICY_RULE_PRIORITY_TYPE_NAMES.includes(instance.elemID.typeName) ? 'rules' : 'policies'
 
-const markRuleAsWeakReference = async (instance: InstanceElement): Promise<ReferenceInfo[]> => {
+const markInstancesAsWeakReference = async (instance: InstanceElement): Promise<ReferenceInfo[]> => {
   const { priorities } = instance.value
   if (priorities === undefined || !Array.isArray(priorities)) {
-    // priorities can be undefined if the policy has no custom rules
+    // priorities can be undefined if their are not custom instances (policies or rules)
```
Comment (edenhassid) [inline packages/okta-adapter/src/weak_references/policy_priorities.ts:38]:
```suggestion
    // priorities can be undefined if they are not custom instances (policies or rules)
```

File: packages/okta-adapter/src/weak_references/policy_priorities.ts [hunk 12]
```diff
@@ -23,53 +23,59 @@ import {
...
     return []
   }
 
   return awu(priorities)
-    .map(async (rule, index) =>
-      isReferenceExpression(rule)
-        ? { source: instance.elemID.createNestedID(index.toString()), target: rule.elemID, type: 'weak' as const }
+    .map(async (ref, index) =>
+      isReferenceExpression(ref)
+        ? { source: instance.elemID.createNestedID(index.toString()), target: ref.elemID, type: 'weak' as const }
```
Comment (edenhassid) [inline packages/okta-adapter/src/weak_references/policy_priorities.ts:48]: i think you should also add the field `priorities` in the `createNestedID`

**Review by edenhassid — COMMENTED** [review]
looks good! 🔥 added some small comments :)

File: packages/okta-adapter/src/dependency_changers/policy_and_rules_to_priority.ts [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+import { ALL_SUPPORTED_POLICY_NAMES, ALL_SUPPORTED_POLICY_RULE_NAMES } from '../filters/policy_priority'
+
+const createDependencyChange = (
+  policyOrPolicyRule: deployment.dependency.ChangeWithKey<ModificationChange<InstanceElement>>,
+  priority: deployment.dependency.ChangeWithKey<ModificationChange<InstanceElement>>,
+): DependencyChange[] => [dependencyChange('add', priority.key, policyOrPolicyRule.key)]
+
+/*
+ * This dependency changer is used to add a dependency from policy or policyRule priority instances
+ * because we need the policy/policyRule to be deployed before its priority.
```
[the remark this answers is not in the document]
Comment (idoamit8) [reply packages/okta-adapter/src/dependency_changers/policy_and_rules_to_priority.ts:39]: there is a parent child relationship between those types. I need to add the dependency only for modifications.

File: packages/okta-adapter/src/dependency_changers/policy_and_rules_to_priority.ts [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+  const [policiesAndRulesChanges, priorityChanges] = _.partition(
+    modificationInstanceChanges.filter(change =>
+      [
+        ...ALL_SUPPORTED_POLICY_NAMES,
+        ...ALL_SUPPORTED_POLICY_RULE_NAMES,
+        ...POLICY_PRIORITY_TYPE_NAMES,
+        ...POLICY_RULE_PRIORITY_TYPE_NAMES,
+      ].includes(getChangeData(change.change).elemID.typeName),
+    ),
```
[the remark this answers is not in the document]
Comment (idoamit8) [reply packages/okta-adapter/src/dependency_changers/policy_and_rules_to_priority.ts:57]: I think it does do it before the actual partition.

File: packages/okta-adapter/src/dependency_changers/policy_and_rules_to_priority.ts [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+        getChangeData(change.change).elemID.typeName,
+      ),
+  )
+
+  if (_.isEmpty(policiesAndRulesChanges) || _.isEmpty(priorityChanges)) {
+    return []
+  }
+
+  return policiesAndRulesChanges.flatMap(policyOrRuleChange =>
+    priorityChanges.map(priorityChange => createDependencyChange(policyOrRuleChange, priorityChange)).flat(),
```
[the remark this answers is not in the document]
Comment (idoamit8) [reply packages/okta-adapter/src/dependency_changers/policy_and_rules_to_priority.ts:69]: The answer is yes. 
But lets say a,b,c are policy rules and A,B,C are their priorities. 
I actually need a -> A, b-> B etc but I am doing a -> A, a-> B, a -> c, b->A, b-> B etc.

[the remark this answers is not in the document]
Comment (edenhassid) [reply packages/okta-adapter/src/dependency_changers/policy_and_rules_to_priority.ts:69]: why is this ok? that you add more dependencies than needed?

File: packages/okta-adapter/src/dependency_changers/policy_and_rules_to_priority.ts [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+      ),
+  )
+
+  if (_.isEmpty(policiesAndRulesChanges) || _.isEmpty(priorityChanges)) {
+    return []
+  }
+  const policyOrRuleNametoChange = Object.fromEntries(
+    policiesAndRulesChanges.map(policyOrRuleChange => {
+      const instance = getChangeData(policyOrRuleChange.change)
+      return [instance.elemID.getFullName(), policyOrRuleChange]
```
Comment (edenhassid) [inline packages/okta-adapter/src/dependency_changers/policy_and_rules_to_priority.ts:75]: nit: ```suggestion
  const policyOrRuleNametoChange = _.keyBy(policiesAndRulesChanges, policyOrRuleChange => getChangeData(policyOrRuleChange.change).elemID.getFullName))
```

**Review by edenhassid — APPROVED** [review]
LGTM! 🔥 added a small nit :)

**Pull request merged**